### PR TITLE
BUG: Add test to ensure, that bug will not occur again. #33058

### LINKED ISSUE #33058: Indexing broken inside groupby - apply
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import pdb

df = pd.DataFrame(
	{
		'col1': ['A', 'A', 'A', 'B', 'B', 'B'],
		'col2': [1, 2, 3, 4, 5, 6],
	}
)

def fn(x):
	pdb.set_trace()
	x.col2[x.index[-1]] = 0
	return x.col2

result = df.groupby(['col1'], as_index=False).apply(fn)
print(result)

```
#### Problem description
The expected output is:
```
0  0    1
   1    2
   2    0
1  3    4
   4    5
   5    0
```
Instead, I get a Series one row longer than expected:
```
0  0    1
   1    2
   2    0
1  3    4
   4    5
   5    6
   5    0
```
The problem seems to come from processing the second group (col1 == 'B'), where indices do not match row numbers. If I stand at the breakpoint (pdb.set_trace()), I can run this with the following results:
```
-> x.col2[x.index[-1]] = 0
(Pdb) x.col2     
3    4
4    5
5    6
Name: col2, dtype: int64
(Pdb) x.col2[5]
*** KeyError: 5
(Pdb) x.col2[5] = 0
(Pdb) x.col2
3    4
4    5
5    6
5    0
Name: col2, dtype: int64
(Pdb) x.col2[5]
5    6
5    0
Name: col2, dtype: int64
(Pdb) x.col2[5] = 0
(Pdb) x.col2
3    4
4    5
5    0
5    0
Name: col2, dtype: int64
```

#### Expected output
```
0  0    1
   1    2
   2    0
1  3    4
   4    5
   5    0
```
This was working before. Unfortunately, I do not know what Pandas version it was.

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.4.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.5.13-050513-generic
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.0.3
numpy            : 1.17.2
pytz             : 2019.3
dateutil         : 2.8.0
pip              : 19.2.3
setuptools       : 40.6.2
Cython           : 0.29.13
pytest           : 5.2.1
hypothesis       : None
sphinx           : 2.2.0
blosc            : None
feather          : None
xlsxwriter       : 1.2.1
lxml.etree       : 4.4.1
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.8.0
pandas_datareader: None
bs4              : 4.8.0
bottleneck       : 1.2.1
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.4.1
matplotlib       : 3.1.1
numexpr          : 2.7.0
odfpy            : None
openpyxl         : 3.0.0
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : 5.2.1
pyxlsb           : None
s3fs             : None
scipy            : 1.3.1
sqlalchemy       : 1.3.9
tables           : 3.5.2
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : 1.3.0
xlsxwriter       : 1.2.1
numba            : 0.45.1

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] xref #33058
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

Should I add a whatsnew entry? It's not really new, if it is backported to 1.0.1.

### CODE REVIEW

**Comment by simonjayhawkins** [comment]
Thanks @phofl for the PR. changing closes #33058 to xref until a decision re backport.

> Should I add a whatsnew entry?

not needed

File: pandas/tests/groupby/test_apply.py [hunk 1]
```diff
@@ -866,3 +866,24 @@ def fct(group):
         [[1.0, 2.0], [3.0], [np.nan]], index=pd.Index(["a", "b", "none"], name="A")
     )
     tm.assert_series_equal(result, expected)
+
+
+def test_apply_function_adds_a_row():
```
Comment (simonjayhawkins) [inline pandas/tests/groupby/test_apply.py]:
```suggestion
def test_apply_function_with_indexing():
```

**Review by simonjayhawkins — APPROVED** [review]
Thanks @phofl lgtm pending green

**Comment by simonjayhawkins** [comment]
marked as 1.1 but could backport if fixed on 1.0.x

File: pandas/tests/groupby/test_apply.py (hunk 1 above)
Comment (phofl) [reply pandas/tests/groupby/test_apply.py]: Changed the test name

**Comment by simonjayhawkins** [comment]
@phofl can you merge upstream master

**Comment by phofl** [comment]
@simonjayhawkins Done.

**Review by simonjayhawkins — APPROVED** [review]
@phofl lgtm @jorisvandenbossche

**Review by WillAyd — CHANGES_REQUESTED** [review]
So I'm not sure this is worth adding a test for - mutating the argument during a `.apply` seems fraught with peril and I'm a little hesitant to make guarantees about behavior

@jorisvandenbossche do you have a strong preference towards adding this?

**Comment by jreback** [comment]
> So I'm not sure this is worth adding a test for - mutating the argument during a `.apply` seems fraught with peril and I'm a little hesitant to make guarantees about behavior
> 
> @jorisvandenbossche do you have a strong preference towards adding this?

this is fine, we currently support mutating inside an apply, though we have an issue to deprecate.

File: pandas/tests/groupby/test_apply.py [hunk 2]
```diff
@@ -880,3 +880,24 @@ def test_apply_function_index_return(function):
         index=pd.Index([1, 2, 3], name="id"),
     )
     tm.assert_series_equal(result, expected)
+
+
```
Comment (jreback) [inline pandas/tests/groupby/test_apply.py:884]: can you put this in a *new* file, call it ``test_apply_mutate.py``; there are 2 tests 

```
pandas/tests/groupby/test_groupby.py:def test_mutate_groups():
pandas/tests/groupby/test_groupby.py:def test_no_mutate_but_looks_like():
```

that should also be moved there.

**Comment by WillAyd** [comment]
sounds good

**Review by jreback — CHANGES_REQUESTED** [review]
see my comments & ping on green.

**Comment by WillAyd** [comment]
Thanks @phofl

**Comment by WillAyd** [comment]
Sorry @jreback merged while you were commented; i can do in a follow up

**Comment by jreback** [comment]
thanks @WillAyd

**Pull request merged**